targets/TARGET_Atmel/TARGET_SAM_CortexM0P/flash_api.c

```diff
@@ -0,0 +1,172 @@
+ * @param size The number of bytes to read
+ * @return 0 for success, -1 for error
+ */
+int32_t flash_read(flash_t *obj, uint32_t address, uint8_t *data, uint32_t size) {
+    // apparently you can read / write multiple pages...
+    uint32_t page_size = flash_get_page_size(obj);
+    if (size % page_size != 0) return -1;
+
+    int32_t bytes_to_read = (int32_t)size;
+    uint8_t *data_slice = data;
+    while (bytes_to_read > 0) {
+        enum status_code status = nvm_read_buffer(address, data_slice, page_size);
+
+        if (status != STATUS_OK) {
+            return -1;
+        }
+
+        data_slice += page_size;
+        bytes_to_read -= page_size;
+    }
+
+    if (bytes_to_read != 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+/** Program pages starting at defined address
+ *
+ * The pages should not cross multiple sectors.
+ * This function does not do any check for address alignments or if size is aligned to a page size.
+ * @param obj The flash object
+ * @param address The sector starting address
+ * @param data The data buffer to be programmed
+ * @param size The number of bytes to program
+ * @return 0 for success, -1 for error
+ */
+int32_t flash_program_page(flash_t *obj, uint32_t address, const uint8_t *data, uint32_t size) {
+    // apparently you can read / write multiple pages...
+    uint32_t page_size = flash_get_page_size(obj);
+    if (size % page_size != 0) return -1;
+
+    int32_t bytes_to_write = (int32_t)size;
+    uint8_t *data_slice = data;
+    while (bytes_to_write > 0) {
+        enum status_code status = nvm_write_buffer(address, data_slice, page_size);
+
+        if (status != STATUS_OK) {
+            return -1;
+        }
+
+        data_slice += page_size;
+        bytes_to_write -= page_size;
+    }
+
+    if (bytes_to_write != 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+/** Get sector size
+ *
+ * @param obj The flash object
+ * @param address The sector starting address
+ * @return The size of a sector
+ */
+uint32_t flash_get_sector_size(const flash_t *obj, uint32_t address) {
+    if (address < flash_get_start_address(obj)) {
+        return MBED_FLASH_INVALID_SIZE;
+    }
+
+    if (address >= flash_get_start_address(obj) + flash_get_size(obj)) {
+        return MBED_FLASH_INVALID_SIZE;
+    }
+
+    return obj->params.page_size * NVMCTRL_ROW_PAGES;
+}
+
+/** Get page size
+ *
+ * The page size defines the writable page size
+ * @param obj The flash object
+ * @return The size of a page
+ */
+uint32_t flash_get_page_size(const flash_t *obj) {
+    return obj->params.page_size;
+}
+
+/** Get start address for the flash region
+ *
+ * @param obj The flash object
+ * @return The start address for the flash region
+ */
+uint32_t flash_get_start_address(const flash_t *obj) {
+    return FLASH_ADDR;
+}
+
+/** Get the flash region size
+ *
+ * @param obj The flash object
+ * @return The flash region size
+ */
+uint32_t flash_get_size(const flash_t *obj) {
```

Comment:
other HAL files follow mostly our coding style, `{` new line for functions in this file

```
uint32_t flash_get_size(const flash_t *obj)
{

}

```

Reply:
Done.